feat(pack): use `nvim_echo` with `kind='progress'` to report progress

echasnovski opened the pull request:
Problem: Progress reports use plain `nvim_echo()` with manually
  constructed messages and populate history on every call.

Solution: Use `nvim_echo()` with newly added `kind=progress` which (at
  least for now) is meant to be a unified interface for showing progress
  report. Also save in history only first and last progress report
  messages.

---

While testing this, I noticed that #35533 made message with 0 percent not show any percent at all. For example, it shows `vim.pack: Installing plugins (0/42)` without aligned percent like `vim.pack:   0% Installing plugins (0/42)`.

Not quite a big deal, but this does make progress report jitter a bit. Is this something that is better to change?

---

Regarding #34774. This PR probably resolves it (the part about notifications will be handled in #34771), but I still have a question about the current progress approach. The whole idea of the #34774 (and `vim.ui.progress`) was for users to be able to exactly customize how progress reports are shown (like in dedicated floating windows, etc.), but that is currently impossible to do cleanly, right?

The only way to do that seems to be a whole (yet unstable) `vim.ui_attach()` route, but this requires implementing the whole `ext_messages` part (which is subject to change), right?

There is `Progress` event to display *another* progress report, but it can not stop `nvim_echo` from showing its messages. This results into two progress reports, which is a bit too much.

There is also ui2. Is it planned/tracked to add an easy way to fully customize how progress report is shown?

luukvbaal commented:
> There is Progress event to display another progress report, but it can not stop nvim_echo from showing its messages. This results into two progress reports, which is a bit too much.

See #35535.

> There is also ui2. Is it planned/tracked to add an easy way to fully customize how progress report is shown?

That sounds out of scope. More realistically there will be a way to map message kinds to either the 'cmd' or 'msg' target. Still need to add support for replacing messages with the same ID.

echasnovski commented:
> > There is Progress event to display another progress report, but it can not stop nvim_echo from showing its messages. This results into two progress reports, which is a bit too much.
> 
> See #35535.

I missed that, thanks! [Enabling `set messagesopt+=progress:none`](https://github.com/neovim/neovim/issues/35535#issuecomment-3236432613) seems like a reasonable solution here. Is it okay to add this to the 0.12 milestone?

justinmk commented:
ASAN found a memory leak, cc @shadmansaleh 

```
-------- Running tests from test/functional/plugin/pack_spec.lua
RUN      T4907 vim.pack add() installs only once: 230.00 ms OK
==================== File /home/runner/work/neovim/neovim/build/log/asan.20405 ====================
= 
= =================================================================
= ==20405==ERROR: LeakSanitizer: detected memory leaks
= 
= Direct leak of 192 byte(s) in 1 object(s) allocated from:
=     #0 0x5624ffc851c0 in realloc (/home/runner/work/neovim/neovim/build/bin/nvim+0xd261c0) (BuildId: 9b0d3a7a6b0b134af43cfe5317a92a0bf9af6fdd)
=     #1 0x56250091c5b4 in xrealloc /home/runner/work/neovim/neovim/src/nvim/memory.c:193:15
=     #2 0x5625009424fb in format_progress_message /home/runner/work/neovim/neovim/src/nvim/message.c:317:5
=     #3 0x56250094072a in msg_multihl /home/runner/work/neovim/neovim/src/nvim/message.c:374:34
=     #4 0x5624ffe5e959 in nvim_echo /home/runner/work/neovim/neovim/src/nvim/api/vim.c:826:8
=     #5 0x562500706baa in nlua_api_nvim_echo /home/runner/work/neovim/neovim/build/src/nvim/auto/lua_api_c_bindings.generated.h:5474:32
=     #6 0x5625014bf065 in lj_BC_FUNCC (/home/runner/work/neovim/neovim/build/bin/nvim+0x2560065) (BuildId: 9b0d3a7a6b0b134af43cfe5317a92a0bf9af6fdd)
= 
= Indirect leak of 33 byte(s) in 1 object(s) allocated from:
=     #0 0x5624ffc84da3 in malloc (/home/runner/work/neovim/neovim/build/bin/nvim+0xd25da3) (BuildId: 9b0d3a7a6b0b134af43cfe5317a92a0bf9af6fdd)
=     #1 0x56250091c074 in try_malloc /home/runner/work/neovim/neovim/src/nvim/memory.c:113:15
=     #2 0x56250091c2d4 in xmalloc /home/runner/work/neovim/neovim/src/nvim/memory.c:147:15
=     #3 0x56250091f231 in arena_alloc /home/runner/work/neovim/neovim/src/nvim/memory.c:743:12
=     #4 0x56250091ff1a in arena_allocz /home/runner/work/neovim/neovim/src/nvim/memory.c:810:15
=     #5 0x562500920008 in arena_memdupz /home/runner/work/neovim/neovim/src/nvim/memory.c:817:15
=     #6 0x5624ffe2d30f in copy_string /home/runner/work/neovim/neovim/src/nvim/api/private/helpers.c:635:30
=     #7 0x562500943573 in format_progress_message /home/runner/work/neovim/neovim/src/nvim/message.c:331:7
=     #8 0x56250094072a in msg_multihl /home/runner/work/neovim/neovim/src/nvim/message.c:374:34
=     #9 0x5624ffe5e959 in nvim_echo /home/runner/work/neovim/neovim/src/nvim/api/vim.c:826:8
=     #10 0x562500706baa in nlua_api_nvim_echo /home/runner/work/neovim/neovim/build/src/nvim/auto/lua_api_c_bindings.generated.h:5474:32
=     #11 0x5625014bf065 in lj_BC_FUNCC (/home/runner/work/neovim/neovim/build/bin/nvim+0x2560065) (BuildId: 9b0d3a7a6b0b134af43cfe5317a92a0bf9af6fdd)
```

shadmansaleh commented:
I think we aren't clearing formated-msg when history is false. Will open a pr today/tomorrow.

echasnovski committed:
feat(pack): use `nvim_echo` with `kind='progress'` to report progress

Problem: Progress reports use plain `nvim_echo()` with manually
  constructed messages and populate history on every call.

Solution: Use `nvim_echo()` with newly added `kind=progress` which (at
  least for now) is meant to be a unified interface for showing progress
  report. Also save in history only first and last progress report
  messages.

echasnovski commented:
> Ok fine I'll wait for a rebase

You shall wait no longer.
